src/EVM/SMT.hs

```diff
@@ -362,138 +391,85 @@ declareBufs names = SMT2 ("; buffers" : fmap declareBuf names <> ("; buffer leng
   where
     declareBuf n = "(declare-const " <> n <> " (Array (_ BitVec 256) (_ BitVec 8)))"
     declareLength n = "(define-const " <> n <> "_length" <> " (_ BitVec 256) (bufLength " <> n <> "))"
-    cexvars = CexVars
-      { calldataV = mempty
-      , buffersV = (fmap toLazyText names)
-      , blockContextV = mempty
-      , txContextV = mempty
-      }
+    cexvars = mempty{buffersV = fmap toLazyText names}
+
+
+referencedBufsGo :: Expr a -> [Builder]
+referencedBufsGo = \case
+  AbstractBuf s -> [fromText s]
+  _ -> []
 
 referencedBufs :: Expr a -> [Builder]
-referencedBufs expr = nubOrd (foldExpr go [] expr)
-  where
-    go :: Expr a -> [Builder]
-    go = \case
-      AbstractBuf s -> [fromText s]
-      _ -> []
+referencedBufs expr = nubOrd $ foldExpr referencedBufsGo [] expr
```

Comment:
We can but in a rather ad hoc way: we could declare a typeclass of types that can by folded over a function of type `(forall a . Expr a -> b)` (Note that already existing [foldable class](https://en.wikibooks.org/wiki/Haskell/Foldable) does not quite work, since it's meant for container types like list)

Reply:
Oh, okay, then it's more trouble then it's worth. Thanks for the explanation, I learned something!

Reply:
afaict `referencedBufs` is only used inside of `referencedBufs'`. I think we could just inline it as a helper. The fact that we have both is probably just a bit of leftover cruft.

Reply:
Both are used in `assertProps`, maybe a typeclass would help cleaning up this code a bit. I'm trying this out in my current branch.